### ISSUE TITLE
Sniffnet (installed from .deb) not starting on Debian 11 (stable) due to missing/not available libraries and versions

### ISSUE THREAD

**Issue by sysms** [GyulyVGC/sniffnet#199]
**Describe the bug**
When installed from the current stable .deb (v1.2.0) on up-to-date Debian 11 (current stable), the binary cannot be launched.

**To Reproduce**
Describe steps to reproduce the bug, e.g.:
1. Download current 1.2.0 .deb 
2. Install package via dpkg -i and install dependencies
3. try to start binary sniffnet from commandline
4. Experience error:

sniffnet: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory

**Expected behavior**
Program should be started.

**Desktop (please complete the following information):**
 - OS: Debian Linux 
 - Version 11 (stable)

**Additional context**

ldd /usr/bin/sniffnet | grep "not found"
/usr/bin/sniffnet: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by /usr/bin/sniffnet)
/usr/bin/sniffnet: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /usr/bin/sniffnet)
/usr/bin/sniffnet: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /usr/bin/sniffnet)
/usr/bin/sniffnet: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /usr/bin/sniffnet)
	libssl.so.3 => not found
	libcrypto.so.3 => not found

Available libraries on Debian 11:

ll /lib/x86_64-linux-gnu/libssl*
-rw-r--r-- 1 root root 398K 15. Feb 21:31 /lib/x86_64-linux-gnu/libssl3.so
-rw-r--r-- 1 root root 988K  5. Feb 22:23 /lib/x86_64-linux-gnu/libssl.a
lrwxrwxrwx 1 root root   13  5. Feb 22:23 /lib/x86_64-linux-gnu/libssl.so -> libssl.so.1.1
-rw-r--r-- 1 root root 584K  5. Feb 22:23 /lib/x86_64-linux-gnu/libssl.so.1.1

ll /lib/x86_64-linux-gnu/libcrypto*
-rw-r--r-- 1 root root 5,3M  5. Feb 22:23 /lib/x86_64-linux-gnu/libcrypto.a
lrwxrwxrwx 1 root root   16  5. Feb 22:23 /lib/x86_64-linux-gnu/libcrypto.so -> libcrypto.so.1.1
-rw-r--r-- 1 root root 3,0M  5. Feb 22:23 /lib/x86_64-linux-gnu/libcrypto.so.1.1

**Comment by GyulyVGC** [comment]
I built it on Ubuntu-latest, could this be the problem?

**Comment by GyulyVGC** [comment]
Other folks already pointed out similar issues on Debian, while seems to work fine in Ubuntu.

I'll come back to you when I have found a solution.

In the meantime, it'd be amazing if you could build it with cargo (requires Rust installed).
If in that case it works, it's definitely a problem of compatibility between the Ubuntu Sniffnet binary and the Debian distro.

**Comment by sysms** [comment]
Thank you for looking into this.

> I built it on Ubuntu-latest, could this be the problem?

Yes, this would explain it.

>it's definitely a problem of compatibility between the Ubuntu Sniffnet binary and the Debian distro.

Ubuntu latest has completely different (newer) libraries than Debian (stable).
Also the versioning/naming of libraries may be different.

Welcome to dependency hell on Linux. :D 

So: yes, it's definitely a compatibility problem between the binary compiled on/for Ubuntu, trying to be run on Debian.

Maybe a static compiled binary could be the solution here (if that's possible) - or publishing as a Flatpak or AppImage.

**Comment by GyulyVGC** [comment]
Thanks very much for your confirmation. And yes, managing all the different distros is a nightmare.

@4r7if3x is going to help me package an AppImage / Flatpak in the future, and I will take care of pinging you.

Maybe it's a stupid question but: could Homebrew work for you?

**Comment by sysms** [comment]
>@4r7if3x is going to help me package an AppImage / Flatpak in the future, and I will take care of pinging you.

Thank you very much! Much appreciated!

>could Homebrew work for you?

For me personally not, because I don't want another package manager on my system which I have to take care of.

But maybe others could benefit?

**Comment by 4r7if3x** [comment]
> @4r7if3x is going to help me package an AppImage / Flatpak in the future, and I will take care of pinging you.

Yes. Unfortunately, I've been quite busy lately and haven't been able to spend as much time as I'd like on testing and improving the new packaging feature. But it's definitely next up on my to-do list...

**Comment by GyulyVGC** [comment]
Don't worry man!

**Comment by 4r7if3x** [comment]
Same as #225, the issue is that the `reqwest` crate defaults to using the `native-ssl` feature, which relies on v1.1 of _OpenSSL_. While _Debian_ aligns with this version by default, Ubuntu adopts v3. To resolve this matter, you can opt to build the project with the `rustls-tls` feature, or alternatively, ask the user manually install the older `libssl1.1` package.

P.S. I've already addressed this issue in my incoming PR.

**Comment by onlineapps-cloud** [comment]
hi, on mx-linux not working, when i try to open app, i received console to enter sudo password, after i enter sudo password, nothing happens, app not running :(

**Comment by GyulyVGC** [comment]
> hi, on mx-linux not working, when i try to open app, i received console to enter sudo password, after i enter sudo password, nothing happens, app not running :(

Hi @onlineapps-cloud 
We need more info: 
- did you install the .deb package, right? 
- can you try opening a terminal typing `sudo -E sniffnet` to see if that solves the issue?
- did you install the [required dependencies](https://github.com/GyulyVGC/sniffnet#required-dependencies)?
- can you give us more details about the behaviour of the app? (error messages etc..)

**Comment by RP-11** [comment]
I have the same issue on Linux Mint 20.3 

` sudo -E sniffnet                                                                                                                  127 ↵
sniffnet: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or director`

**Comment by 4r7if3x** [comment]
> I have the same issue on Linux Mint 20.3 `sudo -E sniffnet` `sniffnet: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or director`

You probably don't have libssl1.1 installed on your system. This issue will be resolved in a near future update.

**Comment by RP-11** [comment]
I checked and it seems it is installed

```
$apt list libssl1.1                                                                                                                  1 ↵
Listing... Done
libssl1.1/focal-updates,focal-security,now 1.1.1f-1ubuntu2.19 amd64 [installed]
libssl1.1/focal-updates,focal-security,now 1.1.1f-1ubuntu2.19 i386 [installed]
```

**Comment by 4r7if3x** [comment]
> sniffnet: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or director

You might also want to check if it's actually loaded using `ldd`...

**Comment by RP-11** [comment]
No it is not loaded. this is the full output when I try to start sniffnet and they grep for libssl
```
$ sniffnet                  
sniffnet: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
$ ldd /usr/bin/sniffnet | grep libssl                                                                                                                                                                              
	libssl.so.3 => not found

```

**Comment by ZEROF** [comment]
Right now using cargo is the solution. Small how to install ( tested on Ubuntu 20.04):

Install cargo:

`curl https://sh.rustup.rs -sSf | sh`

Install dependency (one is missing in documentation libfontconfig1-dev)

`sudo apt install libfontconfig1-dev libasound2-dev libfontconfig1 libpcap-dev`

`cargo install sniffnet`

`sudo setcap cap_net_raw,cap_net_admin=eip /home/$USER/.cargo/bin/sniffnet`

Start using:

`sniffnet`

This allow running sniffnet as regular user, not root.

**Comment by GyulyVGC** [comment]
> Right now using cargo is the solution.

Thanks for pointing out.
Using cargo will perform the build step directly on the target machine, therefore in this case all the dependency should actually be on point 👍

**Comment by GyulyVGC** [comment]
Hey @ZEROF @sysms @RP-11 @onlineapps-cloud 

We changed the building strategy and we are looking for feedback/testers to see if we finally solved this issue.

Could you kindly try the new package available below?

[Sniffnet_LinuxDEB_amd64.deb.zip](https://github.com/GyulyVGC/sniffnet/files/11681994/Sniffnet_LinuxDEB_amd64.deb.zip)

Thanks in advance.

**Comment by sysms** [comment]
@GyulyVGC I can confirm, that this DEB (better: it's contents :) ) now work fine on Debian 11 (current stable)!

Sniffnet now starts without errors.

Thank you for your efforts!

**Comment by GyulyVGC** [comment]
Thanks for the confirmation! 

Don't worry about the issue, I will personally close it after other confirmations or eventually due to inactivity.

I'm happy it's working for you for the moment!!

**Comment by ZEROF** [comment]
> Hey @ZEROF @sysms @RP-11 @onlineapps-cloud
> 
> We changed the building strategy and we are looking for feedback/testers to see if we finally solved this issue.
> 
> Could you kindly try the new package available below?
> 
> [Sniffnet_LinuxDEB_amd64.deb.zip](https://github.com/GyulyVGC/sniffnet/files/11681994/Sniffnet_LinuxDEB_amd64.deb.zip)
> 
> Thanks in advance.

Just tested, looks OK for me on BackBox linux, based on Ubuntu. GJ.
